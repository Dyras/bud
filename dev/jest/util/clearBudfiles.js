const {noop} = require('lodash')
const rimraf = require('rimraf')

module.exports = async () => {
  return rimraf(
    `${process.cwd()}/examples/*/.budfiles`,
    {
      glob: true,
    },
    noop,
  )
}
