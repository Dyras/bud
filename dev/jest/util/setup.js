/* eslint-disable no-console */
// @ts-check

const chalk = require('chalk')
const execa = require('execa')

const clearArtifacts = require('./clearArtifacts')
const paths = require('./paths')

const task = async (cmd, paths) => {
  console.log(chalk.blue`\nyarn bud ${cmd}\n`)

  await Promise.all(
    paths.map(async ex => {
      const task = execa.command(`yarn bud ${cmd}`, {
        cwd: ex.cwd,
      })
      task.stdout.pipe(process.stdout)
      await task
    }),
  )
}

module.exports = async () => {
  await clearArtifacts()

  const examples = await paths()

  await task('build --no-dashboard', examples)
}
