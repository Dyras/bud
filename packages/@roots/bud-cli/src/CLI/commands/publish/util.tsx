import {chalk, globby} from '@roots/bud-support'
import {dirname} from 'path'

/**
 * Publishable matches (glob pattern)
 */
export const publishables = [
  ...globby
    .sync([
      '**/node_modules/bud-*/publish/*',
      '**/node_modules/*/bud-*/publish/*',
    ])
    .filter(
      (value, index, self) => self.indexOf(value) === index,
    ),
]

export const formatted = publishables.reduce((a, file) => {
  const ext = dirname(dirname(file)).replace('node_modules/', '')

  const template = file
    .replace(`${ext}/publish/`, '')
    .replace('node_modules/', '')

  return `${a} \n${chalk.blue(ext)} ${chalk.green(template)}`
}, ``)
