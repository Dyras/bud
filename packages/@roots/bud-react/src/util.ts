/**
 * @module @roots/bud-react
 */

import type {Configuration} from 'webpack'

/**
 * @function addRefresh
 */
function addRefresh(entries, [name, assets]) {
  return {
    ...(entries ?? {}),
    [name]: {
      ...assets,
      import: [
        'react-refresh/runtime',
        ...(assets.import ?? []),
      ],
    },
  }
}

/**
 * @function reducer
 */
function reducer(entry: Configuration['entry']) {
  return entry
    ? Object.entries(entry).reduce(addRefresh, entry)
    : {}
}

/**
 * @exports default
 */
export default reducer
