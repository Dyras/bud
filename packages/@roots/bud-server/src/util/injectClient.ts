/**
 * @module @roots/bud-server
 */

import {Framework} from '@roots/bud-framework'
import * as Webpack from 'webpack'

/**
 * Inject webpack entrypoints with
 * client HMR handling script(s).
 */
declare type InjectClient = (
  app: Framework,
  injection: string[],
) => void

/**
 * Injects webpack entrypoints with HMR client scripts.
 *
 * Filters on `webpack.entry`
 */
const injectClient: InjectClient = (app, injection) => {
  const addScript = (entry: Webpack.Entry): Webpack.Entry => ({
    ...(entry
      ? Object.entries(entry).reduce(
          (entries, [name, asset]) => ({
            ...entries,
            [name]: {
              ...asset,
              import: [...(asset.import ?? []), ...injection],
            },
          }),
          {},
        )
      : {
          app: {
            import: ['index.js', ...injection],
          },
        }),
  })

  app.hooks.on('build/entry', addScript)

  app.children.every((_name: string, child: Framework) => {
    child.hooks.on('build/entry', addScript)
  })
}

/**
 * @exports injectClient
 */
export {injectClient}

/**
 * @exports InjectClient
 */
export type {InjectClient}
