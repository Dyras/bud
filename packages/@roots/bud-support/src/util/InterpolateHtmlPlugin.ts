import * as Webpack from 'webpack'

interface RegularExpressionIndex {
  [key: string]: RegExp
}

class InterpolateHtmlPlugin {
  /**
   * @property {string} name
   */
  public name = 'interpolate-html-plugin'

  /**
   * @property {Webpack.WebpackPluginInstance} htmlWebpackPlugin
   */
  public htmlWebpackPlugin: Webpack.WebpackPluginInstance

  /**
   * @property {RegularExpressionIndex} replacements
   */
  public replacements: RegularExpressionIndex

  /**
   * @constructor
   */
  public constructor(
    htmlWebpackPlugin: Webpack.WebpackPluginInstance,
    replacements: RegularExpressionIndex,
  ) {
    this.htmlWebpackPlugin = htmlWebpackPlugin
    this.replacements = replacements
  }

  /**
   * @function escapeRegExp
   */
  public escapeRegExp(string: String) {
    return string
      .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
      .replace(/-/g, '\\x2d')
  }

  /**
   * @function apply
   * @implements Webpack.WebpackPluginInstance['apply']
   */
  public apply(compiler: Webpack.Compiler): void {
    /**
     * @function modifyHtmlWebpackPluginOptions
     */
    const modifyHtmlWebpackPluginOptions = (
      compilation: any,
    ) => {
      this.htmlWebpackPlugin
        .getHooks(compilation)
        .afterTemplateExecution.tap(
          'InterpolateHtmlPlugin',
          (data: any) => {
            Object.entries(this.replacements).forEach(
              ([key, value]) => {
                data.html = data.html.replace(
                  new RegExp(`%${this.escapeRegExp(key)}%`, 'g'),
                  value,
                )
              },
            )
          },
        )
    }

    compiler.hooks.compilation.tap(
      'InterpolateHtmlPlugin',
      modifyHtmlWebpackPluginOptions,
    )
  }
}

export {InterpolateHtmlPlugin}
