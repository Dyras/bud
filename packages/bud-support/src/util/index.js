"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterpolateHtmlPlugin = exports.formatWebpackMessages = exports.eslintFormatter = exports.processHandler = exports.format = exports.dump = exports.notify = void 0;
var notify_1 = require("./notify");
Object.defineProperty(exports, "notify", { enumerable: true, get: function () { return notify_1.notify; } });
var dump_1 = require("./dump");
Object.defineProperty(exports, "dump", { enumerable: true, get: function () { return dump_1.dump; } });
var format_1 = require("./format");
Object.defineProperty(exports, "format", { enumerable: true, get: function () { return format_1.format; } });
var processHandler_1 = require("./processHandler");
Object.defineProperty(exports, "processHandler", { enumerable: true, get: function () { return processHandler_1.processHandler; } });
const eslintFormatter_1 = __importDefault(require("./eslintFormatter"));
exports.eslintFormatter = eslintFormatter_1.default;
const InterpolateHtmlPlugin_1 = __importDefault(require("./InterpolateHtmlPlugin"));
exports.InterpolateHtmlPlugin = InterpolateHtmlPlugin_1.default;
const formatWebpackMessages_1 = __importDefault(require("./formatWebpackMessages"));
exports.formatWebpackMessages = formatWebpackMessages_1.default;
//# sourceMappingURL=index.js.map