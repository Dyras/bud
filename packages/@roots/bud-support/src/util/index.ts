export {notify} from './notify'
export {dump} from './dump'
export {format} from './format'
export {processHandler} from './processHandler'

import InterpolateHtmlPlugin from './InterpolateHtmlPlugin'
export {InterpolateHtmlPlugin}
