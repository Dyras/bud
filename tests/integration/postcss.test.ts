import {Project} from '../util/integration'

jest.setTimeout(60000)

describe('examples/postcss', () => {
  let project: Project

  beforeAll(async () => {
    project = new Project({
      name: 'postcss',
      dir: 'examples/postcss',
    })

    await project.setup()
  })

  describe('package.json', () => {
    it('matches snapshot', () => {
      expect(project.packageJson).toMatchSnapshot({
        "browserslist": {
          "development": [
            "last 1 chrome version",
            "last 1 firefox version",
            "last 1 safari version",
          ],
          "production": [
            ">0.5%",
            "not dead",
            "not op_mini all",
          ],
        },
        "devDependencies": {
          "@roots/bud": "workspace:*",
          "@roots/bud-babel": "workspace:*",
          "@roots/bud-postcss": "workspace:*",
          "postcss": expect.any(String),
          "postcss-import": expect.any(String),
          "postcss-preset-env": expect.any(String),
        },
        "name": "example-postcss",
        "private": true,
      })
    })
  })

  describe('main.css', () => {
    it('has contents', () => {
      expect(project.assets['app.css'].length).toBeGreaterThan(
        10,
      )
    })

    it('is transpiled', () => {
      expect(
        project.assets['app.css'].includes('@import'),
      ).toBeFalsy()
    })

    it('successfully used @import', () => {
      expect(
        project.assets['app.css'].includes('h2'),
      ).toBeTruthy()
    })
  })
})
