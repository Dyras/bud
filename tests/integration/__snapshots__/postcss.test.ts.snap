// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`examples/postcss package.json matches snapshot 1`] = `
Object {
  "browserslist": Object {
    "development": Array [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version",
    ],
    "production": Array [
      ">0.5%",
      "not dead",
      "not op_mini all",
    ],
  },
  "devDependencies": Object {
    "@roots/bud": "workspace:*",
    "@roots/bud-babel": "workspace:*",
    "@roots/bud-postcss": "workspace:*",
    "postcss": "8.3.6",
    "postcss-import": "14.0.2",
    "postcss-preset-env": "6.7.0",
  },
  "name": "example-postcss",
  "private": true,
}
`;
